### ISSUE TITLE
Inconsistent predictions between CPU and GPU is back on r1.3

### ISSUE THREAD

**Issue by markvdw** [GPflow/GPflow#478]
Hi all, I just wanted to warn that I've found an issue with `predict_y`, similar to the one we saw earlier. On the GPU, on some platforms, I get incorrect results. I haven't looked into what is causing this yet, or whether it's affecting anything else. I'm using the release candidate of r1.3.

Funnily enough, it's only affecting me when using K80s, not other ones. But it could also be a software issue.

Just wanted to flag this just in case anyone else encounters it.

**Comment by alexggmatthews** [comment]
Seems unlikely to be us doesn't it? Maybe report a minimal example on tf main? Could even be a CUDA bug I guess.

**Comment by markvdw** [comment]
It feels similar to #415, which also wasn't our fault directly. That was resolved by an upgrade in TF. Perhaps this is similar.

I don't have time to look into this more deeply at the moment, but if I find some time I'll try to pin it down.

**Comment by alexggmatthews** [comment]
Thanks for taking the time to report it.

**Comment by icouckuy** [comment]
I believe TF 1.3 is the first release that has cholesky gradient on the gpu (#213)? Could this be related?

**Comment by javdrher** [comment]
Think that was already in TF 1.2. However, the gradient for the cholesky op is now different, by default it doesn't seem to be choleskygrad anymore.

**Comment by markvdw** [comment]
I doubt it's the gradients, although I haven't pinned it down enough yet. Last time it was a pretty frustrating exercise. All my models seem to be training as expected. It's only when doing benchmarking that _sometimes_ I get results that are no better than random. This is fixed when I run my code on a different machine with a subtly different software setup.

**Comment by alexggmatthews** [comment]
Re: possible sources of GPU non-determinism. So I've mentioned on here before the under publicised fact that TensorFlow is non-deterministic on GPU:

https://github.com/tensorflow/tensorflow/issues/2732

It seems to be reductions that are associative. Floating point summation is non-associative so the order matters and they don't guarantee the order.

**Comment by alexggmatthews** [comment]
Or is it just the machine that changes the results?

**Comment by markvdw** [comment]
It's both non-deterministic and machine dependent.

**Comment by icouckuy** [comment]
Is it a large difference and are the gpu cards from the same generation? Small numerical precision differences between gpus in combination with non-determinism in reduce operations seems to evidently point to small errors being blow up I guess. Could be a CUDA bug or big enough floating point unit changes between GPU generations.

I was also thinking about the choleskygrad that was changed, but this is only on the cpu (there was no gpu op at the time).

**Comment by alexggmatthews** [comment]
The inconsistencies I mentioned in TF GPU are small. There would have to be some instability in how they are being used for them to become large.

**Comment by awav** [comment]
[offtopic] Here is funny examples which also add fuel to the fire.

```python
In [80]: np_x = np.full(shape=(10, 100, 10000), fill_value=0.12321, dtype=np.float32)
In [81]: x = tf.constant(0.12321, shape=[10, 100, 10000], dtype=tf.float32)
In [82]: np_x.mean()
Out[82]: 0.12320976
In [83]: sess.run(tf.reduce_mean(x))
Out[83]: 0.12479459
In [84]: sess.run(tf.reduce_mean(tf.reduce_mean(tf.reduce_mean(x, axis=2), axis=1)))
Out[84]: 0.12320971
```

**Comment by markvdw** [comment]
There has been discussion about this on the TF github page (possibly linked to above). They are aware of the non-determanism.

I do believe this to be different to the issue I was experiencing, which has since disappeared. I'm going to keep my eye out for what causes this.
